File: mailblastr-go/receiving.go
package mailblastr

import (
	"context"
	"net/http"
	"net/url"
	"strconv"
)

// ReceivedAttachment describes an attachment of a received (inbound) email.
type ReceivedAttachment struct {
	Id                 string `json:"id,omitempty"`
	Filename           string `json:"filename"`
	ContentType        string `json:"content_type"`
	Size               int64  `json:"size"`
	ContentId          string `json:"content_id,omitempty"`
	ContentDisposition string `json:"content_disposition,omitempty"`
	Downloadable       bool   `json:"downloadable"`
	DownloadUrl        string `json:"download_url,omitempty"`
	ExpiresAt          string `json:"expires_at,omitempty"`
	// Url is a backward-compatible relative-path alias.
	Url string `json:"url,omitempty"`
}

// ReceivedEmailRaw points at the original RFC822 message download.
type ReceivedEmailRaw struct {
	DownloadUrl string `json:"download_url"`
	ExpiresAt   string `json:"expires_at"`
}

// ReceivedEmail is an inbound email as returned by GET /emails/receiving/:id.
type ReceivedEmail struct {
	Object string   `json:"object"`
	Id     string   `json:"id"`
	From   string   `json:"from"`
	To     []string `json:"to"`
	// DomainId is the receiving domain the message arrived on.
	DomainId    string   `json:"domain_id"`
	Cc          []string `json:"cc,omitempty"`
	Bcc         []string `json:"bcc,omitempty"`
	ReceivedFor []string `json:"received_for,omitempty"`
	Subject     string   `json:"subject"`
	Html        string   `json:"html,omitempty"`
	Text        string   `json:"text,omitempty"`
	// Headers are the parsed message headers as a name -> value map.
	Headers   map[string]string `json:"headers,omitempty"`
	MessageId string            `json:"message_id,omitempty"`
	// Category is the AI reply intent (replies only): "interested" |
	// "neutral" | "not_interested".
	Category string `json:"category,omitempty"`
	// ReplyToEmailId is the id of the sent email this message replied to,
	// when the reply was recognised.
	ReplyToEmailId string               `json:"reply_to_email_id,omitempty"`
	Spf            string               `json:"spf,omitempty"`
	Verdicts       map[string]any       `json:"verdicts,omitempty"`
	Attachments    []ReceivedAttachment `json:"attachments,omitempty"`
	RawAvailable   bool                 `json:"raw_available"`
	Raw            *ReceivedEmailRaw    `json:"raw,omitempty"`
	RawUrl         string               `json:"raw_url,omitempty"`
	CreatedAt      string               `json:"created_at"`
}

// ForwardReceivedEmailRequest forwards a received email.
type ForwardReceivedEmailRequest struct {
	// From must be a verified sending address (required by the backend).
	From    string   `json:"from"`
	To      []string `json:"to"`
	Subject string   `json:"subject,omitempty"`
}

// ReplyReceivedEmailRequest replies to a received email's sender, threaded
// into the same conversation (subject defaults to "Re: ...").
type ReplyReceivedEmailRequest struct {
	From    string `json:"from"`
	Html    string `json:"html,omitempty"`
	Text    string `json:"text,omitempty"`
	Subject string `json:"subject,omitempty"`
}

// ListReceivedEmailsRequest lists received emails: cursor pagination plus an
// optional received_for filter.
type ListReceivedEmailsRequest struct {
	Limit  int
	After  string
	Before string
	// ReceivedFor restricts to messages received for this address (matches
	// the received_for recipients).
	ReceivedFor string
}

// ReceivingAddressStats is one row of Receiving.ListAddresses: inbound volume
// per receiving address.
type ReceivingAddressStats struct {
	Address string `json:"address"`
	Total   int    `json:"total"`
	Replies int    `json:"replies"`
	// Interested counts replies the AI classifier tagged "interested".
	Interested     int    `json:"interested"`
	LastReceivedAt string `json:"last_received_at"`
}

// ReceivingService handles inbound (received) email — client.Emails.Receiving.
type ReceivingService struct {
	client *Client
}

// List lists received emails. GET /emails/receiving
// For the received_for filter use ListFiltered.
func (s *ReceivingService) List(params *ListParams) (*ListResponse[ReceivedEmail], error) {
	return s.ListWithContext(context.Background(), params)
}

// ListWithContext lists received emails. GET /emails/receiving
func (s *ReceivingService) ListWithContext(ctx context.Context, params *ListParams) (*ListResponse[ReceivedEmail], error) {
	return request[ListResponse[ReceivedEmail]](ctx, s.client, http.MethodGet, listPath("/emails/receiving", params), nil, nil)
}

// ListFiltered lists received emails with an optional received_for filter.
// GET /emails/receiving
func (s *ReceivingService) ListFiltered(params *ListReceivedEmailsRequest) (*ListResponse[ReceivedEmail], error) {
	return s.ListFilteredWithContext(context.Background(), params)
}

// ListFilteredWithContext lists received emails with an optional received_for
// filter. GET /emails/receiving
func (s *ReceivingService) ListFilteredWithContext(ctx context.Context, params *ListReceivedEmailsRequest) (*ListResponse[ReceivedEmail], error) {
	q := url.Values{}
	if params != nil {
		if params.Limit > 0 {
			q.Set("limit", strconv.Itoa(params.Limit))
		}
		if params.After != "" {
			q.Set("after", params.After)
		}
		if params.Before != "" {
			q.Set("before", params.Before)
		}
		if params.ReceivedFor != "" {
			q.Set("received_for", params.ReceivedFor)
		}
	}
	path := "/emails/receiving"
	if enc := q.Encode(); enc != "" {
		path += "?" + enc
	}
	return request[ListResponse[ReceivedEmail]](ctx, s.client, http.MethodGet, path, nil, nil)
}

// ListAddresses reports inbound volume per receiving address. Not paginated —
// HasMore is always false. GET /emails/receiving/addresses
func (s *ReceivingService) ListAddresses() (*ListResponse[ReceivingAddressStats], error) {
	return s.ListAddressesWithContext(context.Background())
}

// ListAddressesWithContext reports inbound volume per receiving address.
// GET /emails/receiving/addresses
func (s *ReceivingService) ListAddressesWithContext(ctx context.Context) (*ListResponse[ReceivingAddressStats], error) {
	return request[ListResponse[ReceivingAddressStats]](ctx, s.client, http.MethodGet, "/emails/receiving/addresses", nil, nil)
}

// Get retrieves a received email. GET /emails/receiving/:id
func (s *ReceivingService) Get(id string) (*ReceivedEmail, error) {
	return s.GetWithContext(context.Background(), id)
}

// GetWithContext retrieves a received email. GET /emails/receiving/:id
func (s *ReceivingService) GetWithContext(ctx context.Context, id string) (*ReceivedEmail, error) {
	return request[ReceivedEmail](ctx, s.client, http.MethodGet, "/emails/receiving/"+esc(id), nil, nil)
}

// ListAttachments lists a received email's attachments. Passing nil params
// returns every attachment: this endpoint only pages when you supply
// pagination params. GET /emails/receiving/:id/attachments
func (s *ReceivingService) ListAttachments(id string, params *ListParams) (*ListResponse[ReceivedAttachment], error) {
	return s.ListAttachmentsWithContext(context.Background(), id, params)
}

// ListAttachmentsWithContext lists a received email's attachments.
func (s *ReceivingService) ListAttachmentsWithContext(ctx context.Context, id string, params *ListParams) (*ListResponse[ReceivedAttachment], error) {
	return request[ListResponse[ReceivedAttachment]](ctx, s.client, http.MethodGet, listPath("/emails/receiving/"+esc(id)+"/attachments", params), nil, nil)
}

// GetAttachment downloads one attachment of a received email as raw bytes
// (the route streams the binary file, not JSON).
// GET /emails/receiving/:id/attachments/:attachmentId
func (s *ReceivingService) GetAttachment(id, attachmentId string) ([]byte, error) {
	return s.GetAttachmentWithContext(context.Background(), id, attachmentId)
}

// GetAttachmentWithContext downloads one attachment of a received email as raw bytes.
func (s *ReceivingService) GetAttachmentWithContext(ctx context.Context, id, attachmentId string) ([]byte, error) {
	return requestRaw(ctx, s.client, http.MethodGet, "/emails/receiving/"+esc(id)+"/attachments/"+esc(attachmentId))
}

// GetRaw downloads the original RFC822/MIME message as raw bytes.
// GET /emails/receiving/:id/raw
func (s *ReceivingService) GetRaw(id string) ([]byte, error) {
	return s.GetRawWithContext(context.Background(), id)
}

// GetRawWithContext downloads the original RFC822/MIME message as raw bytes.
func (s *ReceivingService) GetRawWithContext(ctx context.Context, id string) ([]byte, error) {
	return requestRaw(ctx, s.client, http.MethodGet, "/emails/receiving/"+esc(id)+"/raw")
}

// Forward forwards a received email. POST /emails/receiving/:id/forward
func (s *ReceivingService) Forward(id string, params *ForwardReceivedEmailRequest) (*CreateEmailResponse, error) {
	return s.ForwardWithContext(context.Background(), id, params)
}

// ForwardWithContext forwards a received email. POST /emails/receiving/:id/forward
func (s *ReceivingService) ForwardWithContext(ctx context.Context, id string, params *ForwardReceivedEmailRequest) (*CreateEmailResponse, error) {
	return request[CreateEmailResponse](ctx, s.client, http.MethodPost, "/emails/receiving/"+esc(id)+"/forward", params, nil)
}

// Reply replies to a received email's sender, threaded into the same
// conversation. POST /emails/receiving/:id/reply
func (s *ReceivingService) Reply(id string, params *ReplyReceivedEmailRequest) (*CreateEmailResponse, error) {
	return s.ReplyWithContext(context.Background(), id, params)
}

// ReplyWithContext replies to a received email's sender.
func (s *ReceivingService) ReplyWithContext(ctx context.Context, id string, params *ReplyReceivedEmailRequest) (*CreateEmailResponse, error) {
	return request[CreateEmailResponse](ctx, s.client, http.MethodPost, "/emails/receiving/"+esc(id)+"/reply", params, nil)
}

// Remove deletes a received email. DELETE /emails/receiving/:id
func (s *ReceivingService) Remove(id string) (*RemovedResponse, error) {
	return s.RemoveWithContext(context.Background(), id)
}

// RemoveWithContext deletes a received email. DELETE /emails/receiving/:id
func (s *ReceivingService) RemoveWithContext(ctx context.Context, id string) (*RemovedResponse, error) {
	return request[RemovedResponse](ctx, s.client, http.MethodDelete, "/emails/receiving/"+esc(id), nil, nil)
}

File: mailblastr-dotnet/src/Mailblastr/Models/Receiving.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Mailblastr;

/// <summary>An attachment of a RECEIVED (inbound) email.</summary>
public class ReceivedAttachment
{
    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("filename")]
    public string? Filename { get; set; }

    [JsonPropertyName("content_type")]
    public string? ContentType { get; set; }

    [JsonPropertyName("size")]
    public long Size { get; set; }

    [JsonPropertyName("content_id")]
    public string? ContentId { get; set; }

    [JsonPropertyName("content_disposition")]
    public string? ContentDisposition { get; set; }

    [JsonPropertyName("downloadable")]
    public bool Downloadable { get; set; }

    [JsonPropertyName("download_url")]
    public string? DownloadUrl { get; set; }

    [JsonPropertyName("expires_at")]
    public string? ExpiresAt { get; set; }

    /// <summary>Backward-compatible relative-path alias.</summary>
    [JsonPropertyName("url")]
    public string? Url { get; set; }
}

/// <summary>Pre-signed pointer to the original RFC822/MIME message of a received email.</summary>
public class ReceivedEmailRaw
{
    [JsonPropertyName("download_url")]
    public string DownloadUrl { get; set; } = null!;

    [JsonPropertyName("expires_at")]
    public string ExpiresAt { get; set; } = null!;
}

/// <summary>An inbound email (GET /emails/receiving/:id).</summary>
public class ReceivedEmail
{
    [JsonPropertyName("object")]
    public string Object { get; set; } = "received_email";

    [JsonPropertyName("id")]
    public string Id { get; set; } = null!;

    [JsonPropertyName("from")]
    public string From { get; set; } = null!;

    [JsonPropertyName("to")]
    public List<string> To { get; set; } = new();

    [JsonPropertyName("cc")]
    public List<string>? Cc { get; set; }

    [JsonPropertyName("bcc")]
    public List<string>? Bcc { get; set; }

    [JsonPropertyName("received_for")]
    public List<string>? ReceivedFor { get; set; }

    [JsonPropertyName("subject")]
    public string? Subject { get; set; }

    [JsonPropertyName("html")]
    public string? HtmlBody { get; set; }

    [JsonPropertyName("text")]
    public string? TextBody { get; set; }

    /// <summary>Parsed message headers as a name → value map.</summary>
    [JsonPropertyName("headers")]
    public Dictionary<string, string>? Headers { get; set; }

    [JsonPropertyName("message_id")]
    public string? MessageId { get; set; }

    /// <summary>AI reply intent (replies only): <c>interested</c> | <c>neutral</c> | <c>not_interested</c>.</summary>
    [JsonPropertyName("category")]
    public string? Category { get; set; }

    /// <summary>The sent email this message replied to, when the reply was recognised.</summary>
    [JsonPropertyName("reply_to_email_id")]
    public string? ReplyToEmailId { get; set; }

    [JsonPropertyName("spf")]
    public string? Spf { get; set; }

    [JsonPropertyName("verdicts")]
    public Dictionary<string, JsonElement>? Verdicts { get; set; }

    [JsonPropertyName("attachments")]
    public List<ReceivedAttachment>? Attachments { get; set; }

    [JsonPropertyName("raw_available")]
    public bool RawAvailable { get; set; }

    [JsonPropertyName("raw")]
    public ReceivedEmailRaw? Raw { get; set; }

    [JsonPropertyName("raw_url")]
    public string? RawUrl { get; set; }

    [JsonPropertyName("created_at")]
    public string CreatedAt { get; set; } = null!;
}

/// <summary>Payload for forwarding a received email (POST /emails/receiving/:id/forward).</summary>
public class ReceivedEmailForwardOptions
{
    /// <summary>A verified sending address to forward from (required by the backend).</summary>
    [JsonPropertyName("from")]
    public string From { get; set; } = null!;

    [JsonPropertyName("to")]
    public EmailAddressList To { get; set; } = new();

    [JsonPropertyName("subject")]
    public string? Subject { get; set; }
}

/// <summary>
/// Options for listing received emails (GET /emails/receiving): cursor
/// pagination plus an optional <c>received_for</c> filter.
/// </summary>
public class ReceivedEmailListOptions
{
    /// <summary>Page size (most endpoints cap at 100).</summary>
    public int? Limit { get; set; }

    /// <summary>Cursor: id of the last item on the previous page.</summary>
    public string? After { get; set; }

    /// <summary>Cursor: id of the first item on the next page.</summary>
    public string? Before { get; set; }

    /// <summary>Only messages received for this address (matches the <c>received_for</c> recipients).</summary>
    public string? ReceivedFor { get; set; }
}

/// <summary>
/// Payload for replying to a received email's sender, threaded into the same
/// conversation (POST /emails/receiving/:id/reply). Subject defaults to <c>Re: …</c>.
/// </summary>
public class ReceivedEmailReplyOptions
{
    [JsonPropertyName("from")]
    public string From { get; set; } = null!;

    [JsonPropertyName("html")]
    public string? HtmlBody { get; set; }

    [JsonPropertyName("text")]
    public string? TextBody { get; set; }

    [JsonPropertyName("subject")]
    public string? Subject { get; set; }
}
